### DISCUSSION TITLE
Geometry Colors and Lights

### DISCUSSION THREAD

**Discussion by paulosensorydata** [google/filament#4472] [Q&A]
Hello everyone. I have been trying to convert my SceneKit development into Filament (at least for Android)

I create a geometry with 400 vertices -> 2166 indices
Successfully right now:

```
val surfaceOrientation = SurfaceOrientation
            .Builder()
            .vertexCount(400)
            .positions(vertexData)
            .triangleCount(722)
            .triangles_uint16(indexData)
            .build()

        val vertexTangents = ByteBuffer.allocate(400 * 4 * floatSize).order(ByteOrder.nativeOrder())
        surfaceOrientation.getQuatsAsFloat(vertexTangents)

        vertexBuffer = VertexBuffer.Builder()
            .bufferCount(3)
            .vertexCount(vertexCount)
            .attribute(VertexBuffer.VertexAttribute.POSITION,0, VertexBuffer.AttributeType.FLOAT3, 0, 12)
            .attribute(VertexBuffer.VertexAttribute.COLOR,1, VertexBuffer.AttributeType.UBYTE4,0,4)
            .attribute(VertexBuffer.VertexAttribute.TANGENTS, 2, VertexBuffer.AttributeType.FLOAT4, 3 * floatSize, 16)
            .normalized(VertexBuffer.VertexAttribute.COLOR)
            .build(modelViewer.engine)

        vertexBuffer.setBufferAt(modelViewer.engine, 0, vertexData)
        vertexBuffer.setBufferAt(modelViewer.engine, 1, vertexColor)
        vertexBuffer.setBufferAt(modelViewer.engine, 2, vertexTangents)
```

However the colors seem to be non-natural. I have an indirect light and a sun-type light there, I can see a bit of the reflection when I rotate my geometry, but it is almost unnoticeable. It seems to have low impact right now for some reason. Probably material-related, right?

Already tried multiple materials but didn't get any better.

![Screenshot_1628491546](https://user-images.githubusercontent.com/71468975/128670531-53351a4e-55b0-4177-afb3-62dbc4d0716d.png)

It is very hard to distinguish my geometry high and lows, as you can see.
I guess my question is: Am I doing everything correctly? Do I only need to focus on material now?

**Comment by romainguy** [comment]
What material are you using?

**Reply by paulosensorydata** [reply]
The one present on the previous screenshot is the "square.filamat" from here: https://cx20.github.io/webgl-test/examples/filament/square/index.html

However I used `matc` to create filamat for several different materials from this list: 

<img width="223" alt="Screenshot 2021-08-09 at 09 26 16" src="https://user-images.githubusercontent.com/71468975/128678353-467664f8-a27f-4466-9434-6015af57def6.png">

However without much success from it. I could set the baseColor property on some, but that seemed to override my dynamic coloring

**Reply by romainguy** [reply]
That material is unlit and won’t react to lights. You need to create your own material that’s both lit and uses vertex colors as the baseColor. You can’t use random materials for your specific use case :) The full documentation is here: https://google.github.io/filament/Materials.html

**Reply by paulosensorydata** [reply]
It makes complete sense! So I should also stop using the `vertexColor` I was using above and use just 2 buffers - position and tangent. Is this correct?

**Reply by romainguy** [reply]
You can use vertex color and assign it to baseColor. Check the document above to find how to query the vertex color.

**Reply by paulosensorydata** [reply]
Thank you for quick and accurate answers, @romainguy!
Did it successfully it seems!

For reference:
```
material {
    name : lit,
    shadingModel : lit,

    blending : opaque,
    doubleSided : true,
    postLightingBlending : add,
    specularAmbientOcclusion : bentNormals,

    requires : [
        color
    ],
    parameters : [
        {
           type : sampler2d,
           name : texture
        },
        {
           type : float,
           name : metallic
        },
        {
            type : float,
            name : roughness
        },
        { type : float, name : clearCoatFactor },
        {
            type : float,
            name : ao
        }
    ],
}

fragment {
    void material(inout MaterialInputs material) {
        prepareMaterial(material);
      
        material.baseColor = getColor();
        material.metallic = materialParams.metallic;
        material.roughness = materialParams.roughness;
        material.clearCoat = materialParams.clearCoatFactor;
        material.ambientOcclusion = materialParams.ao;
    }
}
```

And I was now working with lights. I am currently using `lightroom_14b` ibl indirect light. And this is the result:

![New Project](https://user-images.githubusercontent.com/71468975/129020285-27fde96a-3390-4ccb-91b8-4e65bfccae41.png)

My biggest issue now is that I can't make the geometry surface look better - the highlighted parts are really hard to distinguish (depth-wise). Am I correct when assuming lights are the main issue now?

I tried not to use a indirect light and use just one (or more) "Sun" light, but all the geometry become black - it lost all the colors immediately

**Reply by pixelflinger** [reply]
Make sure you enable ambient occlusion as well, this will possibly help with the geometry being not distinguishable.

**Reply by paulosensorydata** [reply]
I am already using it on the material. I am setting it to `0.9f`. If I set it to almost 0, it becomes a lot darker and not-distinguishable
